R/cal-estimate.R

```diff
@@ -0,0 +1,405 @@
+
+  if (run_model == "logistic_spline") {
+    f_model <- expr(!!truth ~ s(!!estimate, k = 10))
+    init_model <- mgcv::gam(f_model, data = .data, family = "binomial", ...)
+    model <- butcher::butcher(init_model)
+  }
+
+  if (run_model == "glm") {
+    f_model <- expr(!!truth ~ !!estimate)
+    init_model <- glm(f_model, data = .data, family = "binomial", ...)
+    model <- butcher::butcher(init_model)
+  }
+
+  model
+}
+
+#------------------------------ >> Isotonic ------------------------------------
+cal_isoreg_dataframe <- function(.data,
+                                 truth,
+                                 estimate,
+                                 sampled = FALSE,
+                                 ...) {
+  sort_data <- dplyr::arrange(.data, !!estimate)
+
+  if (sampled) {
+    sort_data <- dplyr::slice_sample(
+      .data = sort_data,
+      prop = 1,
+      replace = TRUE
+    )
+  }
+
+  x <- dplyr::pull(sort_data, !!estimate)
+
+  truth <- dplyr::pull(sort_data, {{ truth }})
+  y <- as.integer(as.integer(truth) == 1)
+
+  model <- isoreg(x = x, y = y)
+
+  model_stepfun <- as.stepfun(model, ...)
+
+  dplyr::tibble(
+    .estimate = environment(model_stepfun)$x,
+    .adj_estimate = environment(model_stepfun)$y
+  )
+}
+
+cal_isoreg_boot <- function(.data,
+                            truth,
+                            estimate,
+                            times,
+                            ...) {
+  sample.int(10000, times) %>%
+    purrr::map(
+      ~ boot_iso(
+        .data = .data,
+        truth = {{ truth }},
+        estimate = {{ estimate }},
+        seed = .x
+      )
+    ) %>%
+    boot_iso_cal()
+}
+
+boot_iso <- function(.data, truth, estimate, seed) {
+  withr::with_seed(
+    seed,
+    {
+      cal_isoreg_dataframe(
+        .data = .data,
+        truth = {{ truth }},
+        estimate = estimate,
+        sampled = TRUE
+      )
+    }
+  )
+}
+
+boot_iso_cal <- function(x) {
+  # Creates 1,000 predictions using 0 to 1, which become the calibration
+  new_estimates <- seq(0, 1, by = 0.01)
+
+  new_data <- data.frame(
+    .estimate = new_estimates,
+    .adj_estimate = new_estimates
+  )
+
+  new_probs <- purrr::map(x, cal_get_intervals, new_data, expr(.adj_estimate))
+
+  for (i in seq_along(new_probs)) {
+    x <- new_data
+    x$.adj_estimate <- new_probs[[i]]
+    names(x) <- c(".estimate", paste0(".adj_", i))
+    new_probs[[i]] <- x
+  }
+
+  new_probs %>%
+    purrr::reduce(
+      dplyr::inner_join,
+      by = ".estimate"
+    ) %>%
+    dplyr::mutate(
+      .adj_estimate = rowMeans(dplyr::across(dplyr::contains(".adj_")))
+    ) %>%
+    dplyr::select(.estimate, .adj_estimate)
+}
+
+#-------------------------- Binary Objects -------------------------------------
+
+#' @export
+print.cal_binary <- function(x, ...) {
+  cli::cli_div(theme = list(
+    span.val0 = list(color = "blue"),
+    span.val1 = list(color = "yellow")
+  ))
+  cli::cli_h3("Probability Calibration")
```

Comment:
For all of the print methods, let's tell how many data were used to estimate the calibration. 

Also, let's print the number of unique probability values (only for isotonic regression).